### PR TITLE
Generate hostfxr packages/zips

### PR DESCRIPTION
**Pull request by schellap**
Fixes https://github.com/dotnet/core-setup/issues/94

### CODE REVIEW

**Comment by gkhanna79** [comment]
@ellismg @brthor PTAL

File: packaging/windows/hostfxr/registrykeys.wxs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
```
Comment (ellismg) [inline packaging/windows/hostfxr/registrykeys.wxs]: Not sure we need this file, the registry key should be set by the muxer, I think.

Comment (schellap) [inline packaging/windows/hostfxr/registrykeys.wxs]: For servicing detection of hostfxr?

Comment (schellap) [inline packaging/windows/hostfxr/registrykeys.wxs]: This component has its own version independent of the muxer.

File: build_projects/dotnet-host-build/MsiTargets.cs [hunk 2]
```diff
@@ -132,6 +137,31 @@ public static BuildTargetResult GenerateDotnetSharedHostMsi(BuildTargetContext c
 
         [Target]
         [BuildPlatforms(BuildPlatform.Windows)]
+        public static BuildTargetResult GenerateDotnetHostFxrMsi(BuildTargetContext c)
+        {
+            var hostVersion = c.BuildContext.Get<HostVersion>("HostVersion");
+            var hostFxrMsiVersion = hostVersion.GenerateMsiVersion();
```
Comment (schellap) [inline build_projects/dotnet-host-build/MsiTargets.cs]: Msiversion should use FXR not HostVersion

**Comment by schellap** [comment]
The Windows failures are due to a typo: Missing grave accent.

```
        "$AuthWsxRoot\provider.wxs" `
        "$AuthWsxRoot\registrykeys.wxs" <-- HERE
        $InstallFileswsx | Out-Host
```

File: packaging/windows/hostfxr/registrykeys.wxs (hunk 1 above)
Comment (ellismg) [inline packaging/windows/hostfxr/registrykeys.wxs]: Okay.  Can we kill setting the environment variable for PATH at least?

Comment (schellap) [inline packaging/windows/hostfxr/registrykeys.wxs]: Good catch!

**Comment by schellap** [comment]
@brthor can we get this pair-reviewed and pair-tested tomorrow?

@gkhanna79 this is good to go on my side. Porting to rel has no conflicts and works just fine. I'd like to wait for CR rather than port multiple times.

**Comment by schellap** [comment]
Cc @eerhardt @ericstj moves hostfxr into its own zip and msi.

File: build_projects/dotnet-host-build/CompileTargets.cs [hunk 3]
```diff
@@ -357,8 +357,12 @@ public static BuildTargetResult RestoreLockedCoreHost(BuildTargetContext c)
         public static BuildTargetResult PublishSharedFrameworkAndSharedHost(BuildTargetContext c)
         {
             var outputDir = Dirs.SharedFrameworkPublish;
+            Utils.DeleteDirectory(outputDir);
```
Comment (brthor) [inline build_projects/dotnet-host-build/CompileTargets.cs]: Should you recreate the directory here?

File: build_projects/dotnet-host-build/PackageTargets.cs [hunk 4]
```diff
@@ -104,7 +127,10 @@ public static BuildTargetResult CopyCombinedFrameworkHostLayout(BuildTargetConte
             string sharedHostPublishRoot = c.BuildContext.Get<string>("SharedHostPublishRoot");
             Utils.CopyDirectoryRecursively(sharedHostPublishRoot, combinedRoot);
 
-            c.BuildContext["CombinedFrameworkHostRoot"] = combinedRoot;
+            string hostFxrPublishRoot = c.BuildContext.Get<string>("HostFxrPublishRoot");
+            Utils.CopyDirectoryRecursively(hostFxrPublishRoot, combinedRoot);
```
Comment (brthor) [inline build_projects/dotnet-host-build/PackageTargets.cs]: We have two sets of these utilties apparently. 
Not a big deal but I think it'd be better to standardize on FS.cs

File: build_projects/dotnet-host-build/PackageTargets.cs [hunk 5]
```diff
@@ -118,16 +144,27 @@ public static BuildTargetResult GenerateCompressedFile(BuildTargetContext c)
...
         [Target(nameof(PackageTargets.InitPackage))]
         [BuildPlatforms(BuildPlatform.Unix)]
         public static BuildTargetResult GenerateTarBall(BuildTargetContext c)
         {
-            CreateTarBallFromDirectory(c.BuildContext.Get<string>("CombinedFrameworkHostRoot"), c.BuildContext.Get<string>("CombinedFrameworkHostCompressedFile"));
+            CreateTarBallFromDirectory(
+                c.BuildContext.Get<string>("CombinedMuxerHostFxrFrameworkPublishRoot"), 
+                c.BuildContext.Get<string>("CombinedMuxerHostFxrFrameworkCompressedFile"));
+
+            CreateTarBallFromDirectory(
```
Comment (brthor) [inline build_projects/dotnet-host-build/PackageTargets.cs]: If we're zipping hostfxr, should we zip up the muxer as well?

File: build_projects/dotnet-host-build/PublishTargets.cs [hunk 6]
```diff
@@ -300,26 +328,36 @@ public static BuildTargetResult PublishSharedFrameworkInstallerFileToAzure(Build
...
+        {
+            var version = SharedFrameworkNugetVersion;
+            var archiveFile = c.BuildContext.Get<string>("CombinedMuxerHostFxrFrameworkCompressedFile");
+
+            AzurePublisherTool.PublishArchive(archiveFile, Channel, version);
+            return c.Success();
+        }
+
+        [Target]
+        public static BuildTargetResult PublishHostFxrArchiveToAzure(BuildTargetContext c)
```
Comment (brthor) [inline build_projects/dotnet-host-build/PublishTargets.cs]: Same here should we do this for the muxer?

File: build_projects/dotnet-host-build/PrepareTargets.cs [hunk 7]
```diff
@@ -166,8 +166,9 @@ public static BuildTargetResult ExpectedBuildArtifacts(BuildTargetContext c)
             var hostVersion = c.BuildContext.Get<HostVersion>("HostVersion").LockedHostVersion.ToString();
 
             AddInstallerArtifactToContext(c, "dotnet-host", "SharedHost", hostVersion);
+            AddInstallerArtifactToContext(c, "dotnet-hostfxr", "HostFxr", hostVersion);
```
Comment (brthor) [inline build_projects/dotnet-host-build/PrepareTargets.cs]: Use hostfxrversion here

File: build_projects/dotnet-host-build/DebTargets.cs [hunk 8]
```diff
@@ -54,13 +55,48 @@ public static BuildTargetResult GenerateSharedHostDeb(BuildTargetContext c)
...
+
+            var debCreator = new DebPackageCreator(
+                DotNetCli.Stage0,
+                Dirs.Intermediate,
+                dotnetDebToolPackageSource: Dirs.Packages);
+
+            debCreator.CreateDeb(
+                debianConfigFile,
+                packageName,
+                sharedHostVersion,
```
Comment (brthor) [inline build_projects/dotnet-host-build/DebTargets.cs]: Use hostfxrversion here

File: build_projects/dotnet-host-build/DebTargets.cs [hunk 9]
```diff
@@ -54,13 +55,48 @@ public static BuildTargetResult GenerateSharedHostDeb(BuildTargetContext c)
...
+            var hostFxrVersion = c.BuildContext.Get<HostVersion>("HostVersion").LockedHostFxrVersion.ToString();
+            var inputRoot = c.BuildContext.Get<string>("HostFxrPublishRoot");
+            var debFile = c.BuildContext.Get<string>("HostFxrInstallerFile");
+            var debianConfigFile = Path.Combine(Dirs.DebPackagingConfig, "dotnet-hostfxr-debian_config.json");
+
+            var debianConfigVariables = new Dictionary<string, string>()
+            {
+                { "HOSTFXR_BRAND_NAME", Monikers.HostFxrBrandName },
+                { "SHARED_HOST_DEBIAN_VERSION", sharedHostVersion },
+                { "HOSTFXR_DEBIAN_VERSION", hostFxrVersion },
```
Comment (brthor) [inline build_projects/dotnet-host-build/DebTargets.cs]: we don't need this line, only in the shared framework deb

File: build_projects/dotnet-host-build/PackageTargets.cs (hunk 5 above)
Comment (schellap) [inline build_projects/dotnet-host-build/PackageTargets.cs]: There has been no need so far for a muxer zip. The muxer will be installed with MSI to setup `PATH`. `hostfxr.dll` will be delivered through rapid update hence the zip up.

**Comment by schellap** [comment]
@eerhardt can you pls. give a look, if you get a chance? We hacked this in super quick, and there were few issues like copy paste errors (we were still using muxer versions for hostfxr) in a few places. Ready to go in after that.

**Comment by schellap** [comment]
@dotnet-bot test Ubuntu16.04 x64 Release Build please

**Pull request merged**